File: 2023/day5b.py
import math

with open("input-day5.txt", "r") as f:
    lines = f.readlines()
    seeds = []
    maps = []
    for line in lines:
        if "seeds" in line:
            seeds_and_ranges = list(map(int, line.split(":")[1].strip().split()))
            seeds = seeds_and_ranges[::2]
            ranges = seeds_and_ranges[1::2]
        elif "map" in line:
            maps.append([])
        elif line != "\n":
            dest, src, rang = map(int, line.split())
            maps[-1].append((src, dest, rang))

mappings = []

for mapping in maps:
    intervals = [(-math.inf, math.inf, 0)]
    for src, dest, rang in mapping:
        break_intervals = set()
        for i, interval in enumerate(intervals):
            start, end, change = interval
            if start <= src <= end:
                break_intervals.add(i)
            if start <= src + rang - 1 <= end:
                break_intervals.add(i)
                break
        if len(break_intervals) == 1:
            i = break_intervals.pop()
            start, end, change = intervals[i]
            if start < src:
                intervals.insert(i, (start, src - 1, change))
                i = i + 1
            intervals[i] = (src, src + rang - 1, dest - src + change)
            if end > src + rang - 1:
                intervals.insert(i + 1, (src + rang, end, change))
        elif len(break_intervals) == 2:
            i, j = break_intervals
            start1, end1, change1 = intervals[i]
            start2, end2, change2 = intervals[j]
            f = 0
            if start1 < src:
                intervals.insert(i, (start1, src - 1, change1))
                f = 1
            intervals[i + f] = (src, end1, dest - src + change1)
            intervals[j + f] = (start2, src + rang - 1, dest - src + change2)
            if end2 > src + rang - 1:
                intervals.insert(j + f + 1, (src + rang, end2, change2))
            for k in range(i + f + 1, j + f):
                intervals[k] = (intervals[k][0], intervals[k][1], intervals[k][2] + dest - src)
    mappings.append(intervals)

final_mapping = [(-math.inf, math.inf, 0)]

for mapping in mappings:
    new_final_mapping = []
    for start, end, change in final_mapping:
        true_start = start + change
        true_end = end + change
        sub_final_mapping = []
        for m_start, m_end, m_change in mapping:
            if m_start <= true_start and true_end <= m_end:
                sub_final_mapping.append((start, end, change + m_change))
                break
            if m_start <= true_start <= m_end:
                sub_final_mapping.append((start, m_end - change, change + m_change))
            if true_start < m_start and m_end < true_end:
                sub_final_mapping.append((m_start - change, m_end - change, change + m_change))
            if m_start <= true_end <= m_end:
                sub_final_mapping.append((m_start - change, end, change + m_change))
                break
        if len(sub_final_mapping) == 0:
            new_final_mapping.append((start, end, change))
        else:
            new_final_mapping.extend(sub_final_mapping)
    final_mapping = new_final_mapping

min_location = math.inf
                
for seed, rang in zip(seeds, ranges):
    start = seed
    end = seed + rang - 1
    for f_start, f_end, f_change in final_mapping:
        if f_start <= start and end <= f_end:
            min_location = min(min_location, start + f_change, end + f_change)
            break
        if f_start <= start <= f_end:
            min_location = min(min_location, start + f_change, f_end + f_change)
        if start < f_start and f_end < end:
            min_location = min(min_location, f_start + f_change, f_end + f_change)
        if f_start <= end <= f_end:
            min_location = min(min_location, f_start + f_change, end + f_change)
            break

print(min_location)

File: 2023/day5b.c
#include<stdio.h>
#define NMAX 10
#define IMAX 100
struct Interval {
    long start;
    long end;
    long change;
};

typedef struct Interval Interval;

long find_location(long seed, int n, int intervals_count[], Interval intervals[][IMAX]) {
    long tmp = seed;
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < intervals_count[j]; i++) {
            if (intervals[j][i].start <= tmp && tmp <= intervals[j][i].end) {
                tmp += intervals[j][i].change;
                break;
            }
        }
    }
    return tmp;
}

void print_long_array(int n, long arr[]) {
    for (int i = 0; i < n; i++) {
        printf("%ld ", arr[i]);
    }
    printf("\n");
}

void print_int_array(int n, int arr[]) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    FILE *fp = fopen("intervals.txt", "r");
    int intervals_count[NMAX];
    Interval intervals[NMAX][IMAX];
    long int seeds[NMAX];
    long int ranges[NMAX];
    int seeds_count;
    fscanf(fp, "%d", &seeds_count);
    for (int i = 0; i < seeds_count; i++) {
        fscanf(fp, "%ld %ld", seeds + i, ranges + i);
    }
    //print_long_array(seeds_count, seeds);
    //print_long_array(seeds_count, ranges);
    int n;
    fscanf(fp, "%d", &n);
    for (int i = 0; i < n; i++) {
        fscanf(fp, "%d", &intervals_count[i]);
        for (int j = 0; j < intervals_count[i]; j++) {
            fscanf(fp, "%ld %ld %ld", &intervals[i][j].start, &intervals[i][j].end, &intervals[i][j].change);
        }
    }

    //print_int_array(n, intervals_count);

    fclose(fp);

    long min_location = 9223372036854775807;

    for (int i = 0; i < seeds_count; i++) {
        printf("starting seed %ld with range %ld\n", seeds[i], ranges[i]);
        int percentage = 0;
        for (long j = seeds[i]; j < seeds[i] + ranges[i]; j++) {
            long loc = find_location(j, n, intervals_count, intervals);
            /*
            if (loc < min_location) {
                min_location = loc;
            }
            
            if ((j - seeds[i]) * 100 / ranges[i] > percentage) {
                percentage += 1;
                printf("%d\n", percentage);
            }
            */
        }
        printf("done range: %ld\n", min_location);
    }
    printf("%ld\n", min_location);

}